kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,591 @@
+            end
+        end
+
+        if #answers == 0 then
+            answers.errcode = 101
+            answers.errstr = client_errors[101]
+        end
+    end
+
+    process_answers_fields(self, answers)
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+    --log(tries, "query")
+
+    local key = name .. ":" .. qtype
+    stats_count(self.stats, key, "query")
+
+    local r, err = resolver:new(self.r_opts)
+    if not r then
+        return nil, "failed to instantiate the resolver: " .. err
+    end
+
+    local options = { additional_section = true, qtype = qtype }
+    local answers, err, q_tries = r:query(name, options, {})
+    if r.destroy then
+        r:destroy()
+    end
+
+    if not answers then
+        stats_count(self.stats, key, "query_fail")
+        return nil, "Query to DNS server failed: " .. (err or "unknown")
+    end
+
+    process_answers(self, name, qtype, answers)
+
+    stats_count(self.stats, key, answers.errstr and
+                                 "query_err:" .. answers.errstr or "query_succ")
+
+    --log(tries, answers.errstr or #answers)
+
+    return answers, nil, answers.ttl
+end
+
+
+local function start_stale_update_task(self, key, name, qtype)
+    stats_count(self.stats, key, "stale")
+
+    timer_at(0, function (premature)
+        if premature then return end
+
+        local answers = resolve_query(self, name, qtype, {})
+        if answers and (not answers.errcode or answers.errcode == 3) then
+            self.cache:set(key, { ttl = answers.ttl }, answers)
+            insert_last_type(self.cache, name, qtype)
+        end
+    end)
+end
+
+
+local function resolve_name_type_callback(self, name, qtype, opts, tries)
+    local key = name .. ":" .. qtype
+
+    local ttl, err, answers = self.cache:peek(key, true)
+    if answers and not answers.expired then
+        ttl = (ttl or 0) + self.stale_ttl
+        if ttl > 0 then
+            start_stale_update_task(self, key, name, qtype)
+            answers.expired = true
+            return answers, nil, ttl
+        end
+    end
+
+    if opts.cache_only then
+        return { errcode = 100, errstr = client_errors[100] }, nil, -1
+    end
+
+    return resolve_query(self, name, qtype, tries)
+end
+
+
+local function detect_recursion(opts, key)
+    local rn = opts.resolved_names
+    if not rn then
+        rn = {}
+        opts.resolved_names = rn
+    end
+    local detected = rn[key]
+    -- TODO delete
+    if detected then
+        ngx.log(ngx.ALERT, "detect recursion for name:", key)
+    end
+    rn[key] = true
+    return detected
+end
+
+
+local function resolve_name_type(self, name, qtype, opts, tries)
+    local key = name .. ":" .. qtype
+
+    stats_init(self.stats, key)
+    -- log(tries, key)
+
+    if detect_recursion(opts, key) then
+        stats_count(self.stats, key, "fail_recur")
+        return nil, "recursion detected for name: " .. key
+    end
+
+    local answers, err, hit_level = self.cache:get(key, nil,
```

Comment:
I am not sure it is beneficial here, but it could perhaps be nice to group several queries together (by record name or by record type) using the [mlcache:get_bulk](https://github.com/thibaultcha/lua-resty-mlcache/tree/main#get_bulk) feature. I just wanted to point it out since I rarely see it used by mlcache users, yet it is the most powerful feature of mlcache.

Reply:
I seem to recall the old library was making queries for all record types sequentially; here these queries can be scheduled in parallel with minimal effort; just something to consider. `get_bulk()` is very powerful.

Reply:
I'm not sure if it is convenient to convert the two-layer logic of loops for traversing names and types into the get_bulk() parameter list. I need some time to check and give it a try.

Reply:
Finally, we did not use `get_bulk`, which caused the iterate process to generate a larger table (including two dimensions multiplied by type and name). But still thanks for you advice.